### ISSUE TITLE
Use the nixos-rebuild script

### ISSUE THREAD

**Issue by roberth** [tweag/terraform-nixos#27]
As noted by zimbatm in #25, the deploy script is starting to look like `nixos-rebuild`.
It seems like my use case from #25 _is_ actually subsumed by `nixos-rebuild`, unlike my earlier quick interpretation of the docs. Clearly I didn't find the parenthesized bit from `--target-host`:

> (and no build artifacts will be copied to the local machine)

**challenges**

`nixos-rebuild` does need some bootstrapping. [nixos-rebuild.sh](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/nixos-rebuild.sh) has a build process that substitutes some dependencies into the script. Replicating this is a step back from the current script, which takes dependencies from the environment, allowing it to run without a proper `/nix/store` store. This is useful when deploying from restrictive environments without root access or mount namespace capability. So instead of substituting dependencies from the nix store into the script, we should allow values from the environment.

Note: static nix is not officially supported [yet](https://github.com/NixOS/nixpkgs/pull/70024), but here's a branch that makes it work on x86_64-linux deployer machines. https://github.com/tweag/terraform-nixos/compare/master...hercules-ci:install-static-user-nix?expand=1

**Comment by zimbatm** [comment]
Assuming that nix is installed on the system, the deploy script could also build and invoke nixos-rebuild.sh on the fly. I don't think it's going to add too much to the invocation time.

Static nix is interesting but seems a bit orthogonal.

**Comment by roberth** [comment]
Static nix is what allows us to use nix without a `/nix` directory.

The script wont work if it's not in `/nix/store`.
Also you'd have to build it, which requires that the deployer is a Nix `trusted-user` on multi-user systems.

**Comment by zimbatm** [comment]
While it would be nice if the terraform module also worked on machines that didn't have nix installed on them, isn't that an existing requirement because the `nixos-instantiate.sh` script invokes `nix-instantiate`?

**Comment by roberth** [comment]
Not quite. `nixos-instantiate` has to be available in the environment, but it does not have to be installed into `/nix/store`. This works with static nix.

Building any derivations is not feasible in such a restricted environment, but instantiation works.

**Comment by zimbatm** [comment]
Nice. Does that also work on Darwin or does it depend on having user namespaces available?

**Comment by roberth** [comment]
No namespaces used. I see no reason why it shouldn't work on Darwin.
